earth2studio/data/base.py

```diff
@@ -138,3 +139,113 @@ async def fetch(  # type: ignore[override]
             an array of strings with Earth2Studio variable ids.
         """
         pass
+
+
+@runtime_checkable
+class DataFrameSource(Protocol):
+    """Data frame source interface. Commonly used for sparse sensor data"""
+
+    def __call__(
+        self,
+        time: datetime | list[datetime] | TimeArray,
+        variable: str | list[str] | VariableArray,
+    ) -> pd.DataFrame:
+        """Function to get data.
+
+        Parameters
+        ----------
+        time : datetime | list[datetime] | TimeArray
+            Datetime, list of datetimes or array of np.datetime64 to return data for.
+        variable : str | list[str] | VariableArray
+            String, list of strings or array of strings that refer to variables to
+            return.
+
+        Returns
+        -------
+        pd.DataFrame
+            A pandas data-frame like object where the columns are the requested variable
+            list and rows filtered for the time requested.
```

Comment:
Since columns can also be metadata (not just `variable` in the canonical E2S sense of z500, t2m etc) I think we should document that here and in the user guide

Reply:
And I guess document that the data source lexicon will have to define the meaning of these metadata fields

Reply:
Yeah, this was the idea.
So in some sense we also start standardizing meta data and how things are called. I'll add a little more detail.